File: dart/design_patterns/lib/behavioral/template_method.dart
// Template Method Pattern

// Abstract Template
abstract class HouseBuilder {
  // Template method
  void buildHouse() {
    layFoundation();
    buildWalls();
    buildRoof();
    installDoors();
  }

  void layFoundation();
  void buildWalls();
  void buildRoof();
  void installDoors();
}

// Concrete Builders

class WoodenHouseBuilder extends HouseBuilder {
  @override
  void layFoundation() {
    print('Laying wooden foundation');
  }

  @override
  void buildWalls() {
    print('Building wooden walls');
  }

  @override
  void buildRoof() {
    print('Building wooden roof');
  }

  @override
  void installDoors() {
    print('Installing wooden doors');
  }
}

class BrickHouseBuilder extends HouseBuilder {
  @override
  void layFoundation() {
    print('Laying brick foundation');
  }

  @override
  void buildWalls() {
    print('Building brick walls');
  }

  @override
  void buildRoof() {
    print('Building brick roof');
  }

  @override
  void installDoors() {
    print('Installing brick doors');
  }
}

class GlassHouseBuilder extends HouseBuilder {
  @override
  void layFoundation() {
    print('Laying glass foundation');
  }

  @override
  void buildWalls() {
    print('Building glass walls');
  }

  @override
  void buildRoof() {
    print('Building glass roof');
  }

  @override
  void installDoors() {
    print('Installing glass doors');
  }
}

// Usage
void main() {
  final woodenHouseBuilder = WoodenHouseBuilder();
  woodenHouseBuilder.buildHouse();
}

File: swift/DesignPatterns/Behavioral/TemplateMethod.swift
//
//  TemplateMethod.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 20/03/2026.
//

import Foundation

// MARK: - Abstract Template

protocol HouseBuilder {
    func layFoundation()
    func buildWalls()
    func buildRoof()
    func installDoors()
}

extension HouseBuilder {
    // Template method
    func buildHouse() {
        layFoundation()
        buildWalls()
        buildRoof()
        installDoors()
    }
}

// MARK: - Builders

class WoodenHouseBuilder: HouseBuilder {
    func layFoundation() {
        print("Laying wooden foundation")
    }
    
    func buildWalls() {
        print("Building wooden walls")
    }
    
    func buildRoof() {
        print("Building wooden roof")
    }
    
    func installDoors() {
        print("Installing wooden doors")
    }
}

class BrickHouseBuilder: HouseBuilder {
    func layFoundation() {
        print("Laying brick foundation")
    }
    
    func buildWalls() {
        print("Building brick walls")
    }
    
    func buildRoof() {
        print("Building brick roof")
    }
    
    func installDoors() {
        print("Installing brick doors")
    }
}

class GlassHouseBuilder: HouseBuilder {
    func layFoundation() {
        print("Laying glass foundation")
    }
    
    func buildWalls() {
        print("Building glass walls")
    }
    
    func buildRoof() {
        print("Building glass roof")
    }
    
    func installDoors() {
        print("Installing glass doors")
    }
}

// MARK: - Usage

func templateMethodExample() {
    let woodenHouseBuilder = WoodenHouseBuilder()
    woodenHouseBuilder.buildHouse()
}

